### DISCUSSION TITLE
Flickering issue

### DISCUSSION THREAD

**Discussion by edwinsyarief** [hajimehoshi/ebiten#1707] [Q&A]
Hi All,

I've implementing a simple Entity-Scene-Component structure. But I've one issue, when rendering image, using imagecomponent, the result is flickering. You can see at the project here:
https://github.com/edwinsyarief/supershootergame

Please help me, I'm new to golang also ebiten, maybe there is a peformance problem at my structure?

Thanks a lot.

Best Regards,
Edwin

**Comment by hajimehoshi** [comment]
I tested this project and saw a static text. I didn't see flickering.

**Reply by edwinsyarief** [reply]
Hey @hajimehoshi,

Thanks for respond.

Here is the flickering at my side
![HSIyCvB3On](https://user-images.githubusercontent.com/13018338/125157940-08920d00-e198-11eb-8154-95e127419f47.gif)

Best Regards,
Edwin

**Reply by edwinsyarief** [reply]
There are two entities:
- Full black colored box
- White box with FPS on top of it

**Reply by hajimehoshi** [reply]
![image](https://user-images.githubusercontent.com/16950/125158126-dc7f8780-e1a9-11eb-8f8f-a801ae996144.png)

Am I seeing a different thing?

**Reply by hajimehoshi** [reply]
OK I pulled again and saw the flickerings.

**Reply by hajimehoshi** [reply]
Is `ComponentRender` called every frame?

**Reply by edwinsyarief** [reply]
Hi @hajimehoshi ,

Yes, it is called from Draw function. But it will check every draw frame if component is visible, it will be draw in screen

**Reply by hajimehoshi** [reply]
I suggest you create a minimized case to reproduce the issue by reducing the code in your ECS.

**Reply by edwinsyarief** [reply]
Hi @hajimehoshi ,

I've found the problem, here is the thing happened:
At the loaderscene there are one renderer and two entities, which each entity can contains components.
first entity - has image component 1280x720 (black)
second entity - has image component 320x180 (white)

At game->render() function:
loop renderer in scene --> loop entities in scene --> loop components in entity --> screen drawImage

Is there any other way to optimize so there is no flicker, I have a reason why I made the structure like above. I just port my previous .NET ECS into golang.

I'll appreciate a lot if you have any suggestion how to optimize this.

Thank you.

Best Regards,
Edwin

**Reply by hajimehoshi** [reply]
I have no experience about ECS so I don't have a good insight about this.

Ebiten's screen passed via Draw is cleared every frame, then you have to draw everything you need in one frame at the frame once.
